### PR TITLE
bpo-41194: Convert _ast extension to PEP 489

### PR DESCRIPTION
**Pull request by vstinner**
Convert the _ast extension module to PEP 489 "Multiphase
initialization". Replace the global _ast state with a module state.

https://bugs.python.org/issue41194

### CODE REVIEW

**Comment by vstinner** [comment]
With this PR, get_global_ast_state() is still used by multiple functions. It can introduce minor inconsistencies when the _ast extension module is loaded more than once in the same interpreter. But this PR better isolates the _ast module in subinterpreters.

The PyType_GetModule() (PEP 573) cannot be used in ast_type_init() (tp_init slot).

Even if all AST_Type methods and slots are fixed to use PyType_GetModule(), there are still PyAST_mod2obj() and PyAST_obj2mod() functions which still use get_global_ast_state().

Even if get_global_ast_state() has drawbacks, IMHO this PR is worth it.

@corona10: Would you mind to review it?

cc @DinoV

File: Parser/asdl_c.py [hunk 1]
```diff
@@ -1046,40 +1052,55 @@ def emit_defaults(self, name, fields, depth):
...
-        self.emit("return NULL;", 2)
-        self.emit("}", 1)
         self.emit('astmodulestate *state = get_ast_state(m);', 1)
-        self.emit('', 1)
+        self.emit("", 0)
 
-        self.emit("if (!init_types()) {", 1)
-        self.emit("goto error;", 2)
+        self.emit("if (!init_types(state)) {", 1)
+        self.emit("return 1;", 2)
```
Comment (corona10) [inline Parser/asdl_c.py]: Just question: Why return 1, not -1 ? :)

**Review by corona10 — APPROVED** [review]
Looks good with a simple question :)

File: Parser/asdl_c.py (hunk 1 above)
Comment (vstinner) [reply Parser/asdl_c.py]: Hum, it's a mistake. I prefer to return -1. Fixed.

Thanks for the review.

**Comment by vstinner** [comment]
>  continuous-integration/travis-ci Expected — Waiting for status to be reported 

But https://travis-ci.org/github/python/cpython/builds/704696576 is the build and it completed. I don't get it :-(

* Ran for 18 min 13 sec
* Total time 27 min 27 sec
* 9 minutes ago

**Comment by vstinner** [comment]
I close/reopen my issue to force a new Travis CI build :-(

**Comment by corona10** [comment]
I did same thing on https://github.com/python/cpython/pull/21294 :(

**Comment by corona10** [comment]
re-open :)

File: Parser/asdl_c.py [hunk 2]
```diff
@@ -1358,12 +1373,35 @@ def generate_module_def(f, mod):
...
+static astmodulestate*
+get_global_ast_state(void)
 {
-    return &global_ast_state;
+    _Py_IDENTIFIER(_ast);
+    PyObject *name = _PyUnicode_FromId(&PyId__ast);  // borrowed reference
+    if (name == NULL) {
+        return NULL;
+    }
+    PyObject *module = PyImport_GetModule(name);
```
Comment (encukou) [inline Parser/asdl_c.py:1392]: As noted in https://bugs.python.org/issue41631, this is incorrect :( 
`PyImport_GetModule` looks in `sys.modules`, which is modifiable by the user, and so it's not guaranteed to return `_ast`.

**Pull request merged**